### ISSUE TITLE
Open to PR that makes CustomTimingsJson thread-safe?

### ISSUE THREAD

**Issue by jankurianski** [MiniProfiler/dotnet#488]
I got the following crash using MiniProfiler 4.1.0 + MiniProfilerEF6 + ASP.NET MVC 5.  No storage provider (or rather, just the default in-memory one).  Also using ProfiledDbConnection.

```
    type: System.InvalidOperationException
    message: Collection was modified; enumeration operation may not execute.
    HResult: 80131509
    Source: mscorlib
    siteStackTrace: [
           at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
           at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
           at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeList(JsonWriter writer, IEnumerable values, JsonArrayContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
           at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeDictionary(JsonWriter writer, IDictionary values, JsonDictionaryContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
           at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize(JsonWriter jsonWriter, Object value, Type objectType)
           at Newtonsoft.Json.JsonSerializer.SerializeInternal(JsonWriter jsonWriter, Object value, Type objectType)
           at Newtonsoft.Json.JsonConvert.SerializeObjectInternal(Object value, Type type, JsonSerializer jsonSerializer)
           at StackExchange.Profiling.Internal.ExtensionMethods.ToJson(Object o)
           at ServerProfilerLib.ServerProfiler.TimingToMetric(Timing timing, MiniProfiler profiler, ServerProfilerMetricDto root_metric, ServerProfilerSessionDto session, ServerProfilerSource source)
           at ServerProfilerLib.ServerProfiler.Save(MiniProfiler profiler, ServerProfilerSessionDto session, ServerProfilerSource source)
           at QWeb.Site.MvcApplication.Application_EndRequest(Object sender, EventArgs e)
           at System.Web.HttpApplication.SyncEventExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
           at System.Web.HttpApplication.ExecuteStepImpl(IExecutionStep step)
           at System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously)
    ]
```

I have custom storage code that essentially does what the other storage providers do - flatten the nested Timings and call the properties of the flattened Timings including `CustomTimingsJson` which is where the call to `StackExchange.Profiling.Internal.ExtensionMethods.ToJson` originated (I guess .NET inlines the property so it doesn't show up in the stack).  It does this in the same thread as `Application_EndRequest`.

This crash has only happened once over many thousand requests so it must be a race condition.  Most likely somewhere in my profiled code where code is adding or stopping a custom timing in another thread.  However it is extremely hard to track down and other users may suffer the same random error.

I notice that the addition and removal of custom timings is threadsafe: https://github.com/MiniProfiler/dotnet/blob/ff8b503b68695ae92e1c6ffbd38f58d12f4832d3/src/MiniProfiler.Shared/Timing.cs#L357

Would you be open to receiving a pull request that makes the `Timing.CustomTimingsJson` property getter thread-safe also?  I would make a copy of the dictionary and its lists in a thread-safe way (locking on dictionary and locking each list during the copy), and then serialize the copied dictionary.

All of the storage providers use `CustomTimingsJson` as well so they would benefit from this thread safety.

**Comment by NickCraver** [comment]
Hmmm, the issue with adding locking here is: it doesn't solve the fundamental issue. It sounds like you're trying to serialize it _while things are still being added_. That's not an intended or supported use case. A profiler should be stopped before it gets saved (that's what normally happens in the `.Stop()` and `.StopAsync()` paths).

So let's pop up a level - why are you serializing the profiler before it's done profiling? I'd like to understand the use case better here to provide the best advice on a solution :)

**Comment by jankurianski** [comment]
Thanks for your quick response Nick.  I am calling `.Stop()` on the profiler before saving.  Here's a simplified version of my `Application_EndRequest`:

```
        protected void Application_EndRequest(object sender, EventArgs e)
        {
            var profiler = MiniProfiler.Current;
            if (profiler != null) {
                profiler.Stop();
                ServerProfiler.Save(profiler); // my code to flatten the timings of root and do further processing and serialization
            }
        }
```

I'm not adding/removing any timings afterwards, and the saving is being done on the same ASP.NET thread, which is why I'm confused.  In theory `Application_EndRequest` would only be called after my controller actions have finished their work and disposed of their entity framework contexts and profiled DB connections, so it shouldn't be possible.

I have worked around this for now by copying the `CustomTimings` dictionary with matching locks in my code, before serializing to JSON.  I'll just assume this crash was due to an oddity of my code base and close this issue.  Perhaps if other users report a similar crash with the default storage providers in future then it will become relevant.
